pkg/hs/urbit-king/lib/Urbit/Vere/Ports.hs

```diff
@@ -0,0 +1,239 @@
+module Urbit.Vere.Ports (HasPortControlApi(..),
+                         PortControlApi(..),
+                         buildInactivePorts,
+                         buildNATPorts,
+                         requestPortAccess) where
+
+import Control.Monad.STM (check)
+import Urbit.Prelude
+import Network.NATPMP
+import Data.Time.Clock.POSIX
+import Data.Heap
+import Network.Socket
+
+-- This module deals with ports and port requests. When a component wants to
+-- ensure that it is externally reachable, possibly from outside a NAT, it
+-- makes a request to this module to hole-punch.
+
+class HasPortControlApi a where
+  portControlApiL :: Lens' a PortControlApi
+
+data PortControlApi = PortControlApi
+  { pAddPortRequest :: Word16 -> IO ()
+  , pRemovePortRequest :: Word16 -> IO ()
+  }
+
+-- Builds a Ports struct which does nothing when called.
+buildInactivePorts :: PortControlApi
+buildInactivePorts = PortControlApi noop noop
+  where
+    noop x = pure ()
+
+-- Builds a Ports struct which tries to hole-punch by talking to the NAT
+-- gateway over NAT-PMP.
+buildNATPorts :: (HasLogFunc e)
+              => (Text -> RIO e ())
+              -> RIO e PortControlApi
+buildNATPorts stderr = do
+  q <- newTQueueIO
+  async $ portThread q stderr
+  pure $ PortControlApi (addRequest q) (removeRequest q)
+  where
+    addRequest :: TQueue PortThreadMsg -> Word16 -> IO ()
+    addRequest q port = do
+      resp <- newEmptyTMVarIO
+      atomically $
+        writeTQueue q (PTMInitialRequestOpen port (putTMVar resp True))
+      atomically $ takeTMVar resp
+      pure ()
```

Comment:
Because what you're proposing would wait forever. This first atomically adds something to the queue. The second atomically waits for the stm action passed across async boundaries to fire.

And we do this because we don't want to proceed until we have confirmation that the port is open.